control-plane-operator/controllers/azureprivateendpoint/azureprivateendpoint_controller.go

```diff
@@ -0,0 +1,327 @@
+			return ctrl.Result{}, nil
+		}
+		if err := r.delete(ctx, azurePrivateEndpoint, hcp); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to delete resource: %w", err)
+		}
+
+		if controllerutil.ContainsFinalizer(azurePrivateEndpoint, finalizer) {
+			controllerutil.RemoveFinalizer(azurePrivateEndpoint, finalizer)
+			if err := r.Patch(ctx, azurePrivateEndpoint, client.MergeFrom(originalObj)); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer: %w", err)
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Ensure the azurePrivateEndpoint has a finalizer for cleanup
+	if !controllerutil.ContainsFinalizer(azurePrivateEndpoint, finalizer) {
+		controllerutil.AddFinalizer(azurePrivateEndpoint, finalizer)
+		if err := r.Patch(ctx, azurePrivateEndpoint, client.MergeFrom(originalObj)); err != nil {
+			if apierrors.IsConflict(err) {
+				return ctrl.Result{Requeue: true}, nil
+			}
+			return ctrl.Result{}, fmt.Errorf("failed to add finalizer: %w", err)
+		}
+	}
+
+	// Reconcile the AzurePrivateEndpoint
+	originalObj = azurePrivateEndpoint.DeepCopy()
+	if err := r.reconcileAzurePrivateEndpoint(ctx, azurePrivateEndpoint, hcp); err != nil {
+		meta.SetStatusCondition(&azurePrivateEndpoint.Status.Conditions, metav1.Condition{
+			Type:    string(hyperv1.AzurePrivateEndpointAvailable),
+			Status:  metav1.ConditionFalse,
+			Reason:  hyperv1.AzureErrorReason,
+			Message: err.Error(),
+		})
+		if err := r.Status().Patch(ctx, azurePrivateEndpoint, client.MergeFrom(originalObj)); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		return ctrl.Result{}, err
+	}
+
+	meta.SetStatusCondition(&azurePrivateEndpoint.Status.Conditions, metav1.Condition{
+		Type:    string(hyperv1.AzurePrivateEndpointAvailable),
+		Status:  metav1.ConditionTrue,
+		Reason:  hyperv1.AsExpectedReason,
+		Message: "",
+	})
+	if err := r.Status().Patch(ctx, azurePrivateEndpoint, client.MergeFrom(originalObj)); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	log.Info("reconcilation complete")
+	// always requeue to catch and report out of band changes.
+	return ctrl.Result{RequeueAfter: defaultResync}, nil
+}
+
+func (r *AzurePrivateEndpointReconciler) reconcileAzurePrivateEndpoint(ctx context.Context, azurePrivateEndpoint *hyperv1.AzurePrivateEndpoint, hcp *hyperv1.HostedControlPlane) error {
+	// log, err := logr.FromContext(ctx)
+	// if err != nil {
+	// 	return fmt.Errorf("logger not found: %w", err)
+	// }
+	if hcp.Spec.Platform.Azure == nil {
+		return nil
+	}
+	subscriptionID := hcp.Spec.Platform.Azure.SubscriptionID
+
+	cred, err := r.getAzureCredential(ctx, hcp)
+	if err != nil {
+		return err
+	}
+
+	endpointClient, err := armnetwork.NewPrivateEndpointsClient(subscriptionID, cred, nil)
+	if err != nil {
+		return err
+	}
+
+	response, err := endpointClient.BeginCreateOrUpdate(ctx, azurePrivateEndpoint.Spec.ResourceGroupName, azurePrivateEndpoint.Name, armnetwork.PrivateEndpoint{
```

Comment:
Are we okay creating azure cloud resources in the controller? Maybe there's no way to avoid it here?